src/libmeasurement_kit/ooni/dpi_fragment.cpp

```diff
@@ -0,0 +1,373 @@
+    fd_set rfds, wfds;
+
+    while(1) {
+        logger->info("TOP OF WHILE LOOP\n");
+        FD_ZERO(&rfds);
+        FD_ZERO(&wfds);
+        if (!stop_reading) {
+            FD_SET(sockfd, &rfds);
+            logger->info("we want to read");
+        }
+        if (!nwrite->empty() && !stop_writing) {
+            FD_SET(sockfd, &wfds);
+            logger->info("we want to write");
+        }
+
+        tv.tv_sec = 4;
+        tv.tv_usec = 0;
+
+        logger->info("BLOCKING ON SELECT\n");
+        sel_ret = select(sockfd+1, &rfds, &wfds, NULL, &tv);
+        if (sel_ret < 0) {
+            logger->info("select() failed");
+            break;
+        }
+        if (sel_ret == 0) {
+            logger->info("select() timed out.\n");
+            if (do_ssl) {
+                ssl_filter.update(); //XXX shouldn't need to sprinkle these around so much if i do more thinking
+            }
+            break;
+        }
+        if (FD_ISSET(sockfd, &wfds)) {
+            logger->info("WRITABLE\n");
+
+            //XXX clean this up
+            if (do_fragment) {
+                std::size_t found = nwrite->find(hostname);
+                if (found!=std::string::npos) {
+                    logger->info("$$$$$$ found plaintext hostname; hacking planet $$$$$$$$$");
+                    n = write(sockfd, nwrite->c_str(), found+3);
+                    if (n > 0) {
+                        nwrite->erase(0, n);
+                    }
+                    sleep(0.5);
+                }
+            }
+
+            n = write(sockfd, nwrite->c_str(), nwrite->length());
+            if (n > 0) {
+                nwrite->erase(0, n);
+            }
+        }
+
+        if (FD_ISSET(sockfd, &rfds)) {
+            logger->info("READABLE\n");
+            while(1) {
+                n = read(sockfd, &readbuf, sizeof(readbuf));
+                logger->info("read from sockfd: %d", n);
+                if (read > 0) {
+                    size_t cur_size = nread->length();
+                    nread->resize(cur_size + n);
+                    std::copy(readbuf, readbuf + n, nread->begin() + cur_size);
+                }
+                if (n == 0) {
+                    stop_reading = true;
+                    break;
+                }
+                if (static_cast<size_t>(n) != sizeof(readbuf)) {
+                    break;
+                }
+                if(n < 0) {
+                    break;
+                }
+            }
+        }
+
+        if (do_ssl) {
+            ssl_filter.update();
+        }
+        if (awrite->size() == 0 && nwrite->size() == 0) {
+            stop_writing = true;
+        }
+        if (stop_reading && stop_writing) {
+            break;
+        }
+
+    }
+
+    return *aread;
+}
+
+
+
+
+void dpi_fragment(Settings options, Callback<SharedPtr<report::Entry>> callback,
+                        SharedPtr<Reactor> reactor, SharedPtr<Logger> logger) {
+    reactor->call_in_thread(logger, [=]() {
```

Comment:
I suggest to do `[callback = std::move(callback)]`. This is C++14 and should guarantee that the thread uniquely owns the callback. I remember not doing that was leading to weird behavior sometimes.

Reply:
Done.